### PR TITLE
Improvements to support 3rd party types and native apps

### PR DESCRIPTION
**Pull request by marb2000**
Issue: #

## PR Type
What kind of change does this PR introduce?

This change improves user experience to better support 3rd party WinRT components.

 - Feature -

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: 
- [X] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [ ] Contains **NO** breaking changes

XamlApplication class is now a native WinRT component.

### CODE REVIEW

File: build/build.cake [hunk 1]
```diff
@@ -210,28 +228,23 @@ Task("InheritDoc")
...
-	};
-
-    DotNetCoreBuild(win32Solution, buildSettings);
-});
+    .SetConfiguration("Release")
+    //.WithProperty("GenerateLibraryLayout", "true")
+    .UseToolVersion(MSBuildToolVersion.VS2019)
+    .WithTarget("Pack");
 
+    Information("\nBuild Step");
```
Comment (azchohfi) [inline build/build.cake]: Remove this, or change with "Pack Step".

File: build/build.cake [hunk 2]
```diff
@@ -210,28 +228,23 @@ Task("InheritDoc")
...
-	var buildSettings = new DotNetCoreBuildSettings
-	{
-		Configuration = "Release",
-		MSBuildSettings = msbuildSettings
-	};
-
-    DotNetCoreBuild(win32Solution, buildSettings);
-});
+    .SetConfiguration("Release")
+    //.WithProperty("GenerateLibraryLayout", "true")
```
Comment (azchohfi) [inline build/build.cake]: Remove commented code here.

File: build/build.cake [hunk 3]
```diff
@@ -146,61 +146,79 @@ Task("Build")
...
+    .WithTarget("Build");
+
+    // Build once with normal dependency ordering
+    buildSettings = new MSBuildSettings
+    {
+        PlatformTarget = PlatformTarget.x64,
+        MaxCpuCount = 0,
+    }
+    .SetConfiguration("Release")
+    //.WithProperty("GenerateLibraryLayout", "true")
```
Comment (azchohfi) [inline build/build.cake]: Remove commented code here.

File: build/build.cake [hunk 4]
```diff
@@ -146,61 +146,79 @@ Task("Build")
...
-	
-	buildSettings = new DotNetCoreBuildSettings
-	{
-		Configuration = "Release",
-		MSBuildSettings = msbuildSettings
-	};
-
-    DotNetCoreBuild(win32Solution, buildSettings);
+    .SetConfiguration("Release")
+    //.WithProperty("GenerateLibraryLayout", "true")
```
Comment (azchohfi) [inline build/build.cake]: Remove commented code here.

File: build/build.cake [hunk 5]
```diff
@@ -146,61 +146,79 @@ Task("Build")
...
+    .UseToolVersion(MSBuildToolVersion.VS2019)
+    .WithTarget("Restore");
+
+    buildSettings = new MSBuildSettings
+    {
+        PlatformTarget = PlatformTarget.x64,
+        MaxCpuCount = 0,
+    }
+    .SetConfiguration("Release")
+    //.WithProperty("GenerateLibraryLayout", "true")
```
Comment (azchohfi) [inline build/build.cake]: Remove commented code here.

File: build/build.cake [hunk 6]
```diff
@@ -146,61 +146,79 @@ Task("Build")
...
-	
-    DotNetCoreBuild(win32Solution, buildSettings);
+
+    var buildSettings = new MSBuildSettings
+    {
+        PlatformTarget = PlatformTarget.x86,
+        MaxCpuCount = 0,
+    }
+    .SetConfiguration("Release")
+    //.WithProperty("GenerateLibraryLayout", "true")
```
Comment (azchohfi) [inline build/build.cake]: Remove commented code here.

File: Microsoft.Toolkit.Win32.sln [hunk 7]
```diff
@@ -41,6 +41,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Toolkit.Sample.Fo
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{2B59D500-7A53-452F-B0AF-774E0ED98C16}"
 	ProjectSection(SolutionItems) = preProject
+		build\build.cake = build\build.cake
```
Comment (azchohfi) [inline Microsoft.Toolkit.Win32.sln]: I wouldn't add this. It's not part of the solution per se.

File: Directory.Build.props [hunk 8]
```diff
@@ -23,6 +23,19 @@
     <PackageOutputPath>$(MSBuildThisFileDirectory)bin\nupkg</PackageOutputPath>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <XamlApp-Platform Condition="'$(Platform)' != ''">$(Platform)</XamlApp-Platform>
+    <XamlApp-Platform Condition="'$(Platform)' == 'x86'">Win32</XamlApp-Platform>
+    <XamlApp-Platform Condition="'$(Platform)' == 'AnyCPU'">Win32</XamlApp-Platform>
+    <XamlApp-Platform Condition="'$(Platform)'=='x86'">Win32</XamlApp-Platform>
```
Comment (Austin-Lamb) [inline Directory.Build.props]: Why is $(Platform)=='x86' condition checked twice to set to Win32?

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj [hunk 9]
```diff
@@ -67,7 +68,21 @@
...
 
   <ItemGroup>
-    <ProjectReference Include="..\Microsoft.Toolkit.Win32.UI.XamlHost\Microsoft.Toolkit.Win32.UI.XamlHost.csproj" />
+    <PackageReference Include="Microsoft.VCRTForwarders.140" Version="1.0.0-rc" />
   </ItemGroup>
 
+  <PropertyGroup>
+    <XamlApp-Platform Condition="'$(Platform)'=='x86'">Win32</XamlApp-Platform>
+    <XamlApp-Platform Condition="'$(Platform)'=='x64'">x64</XamlApp-Platform>
+    <XamlApp-Platform Condition="'$(Platform)'=='AnyCPU'">x64</XamlApp-Platform>
```
Comment (Austin-Lamb) [inline Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: Why is this XamlApp-Platform stuff repeated here and in Directory.Build.props?

File: Microsoft.Toolkit.Sample.Forms.App/Program.cs [hunk 10]
```diff
@@ -16,15 +16,10 @@ static class Program
...
-            using (new UI.XamlHost.XamlApplication()
-            {
-                RequestedTheme = WUX.ApplicationTheme.Light,
-            })
-            {
-                Application.EnableVisualStyles();
-                Application.SetCompatibleTextRenderingDefault(true);
-                Application.Run(new Form1());
-            }
+            System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor(typeof(Microsoft.Toolkit.Forms.UI.XamlHost.WindowsXamlHostBase).TypeHandle);
```
Comment (Austin-Lamb) [inline Microsoft.Toolkit.Sample.Forms.App/Program.cs]: Did you ever figure out why this workaround is needed, and who is fixing it in .NET Core?

**Comment by marb2000** [comment]
I think that @Ryland Alaniz<mailto:ryalan@microsoft.com> or @Daniel Zilcsak<mailto:danzil@microsoft.com> has more details about it.

File: Microsoft.Toolkit.Sample.Forms.App/Program.cs (hunk 10 above)
Comment (ocalvo) [reply Microsoft.Toolkit.Sample.Forms.App/Program.cs]: I am following up on a separate email thread.

**Comment by michael-hawker** [comment]
@marb2000 can we get a better summary for this PR of details of what the overall PR changes and accomplishes since there's not a direct issue filed (though there is now [one on WinUI repo](https://github.com/microsoft/microsoft-ui-xaml/issues/798))?

Alex was telling me this breaks using MSIX though which seems like a big blocker as they've had a big uptick since //Build.

**Comment by azchohfi** [comment]
I retested today and that issue has been addressed with one of the latest commits. We are good to merge this.

**Pull request merged**